### ISSUE TITLE
Structure for Services and Addons

### ISSUE THREAD

**Issue by stuartpb** [plushu/plushu-services#2]
## Plugin architecture

The plushu-services plugin installs commands and hooks for managing services.

The plushu-addons plugin installs commands for managing instances of these services attached to apps, at an abstract/general level (ie. handling the data that specifies which apps have addons for which services and probably handling the filesystem for them).

Addon implementation plugins like plushu-addons-docker handle the actual rollout/runtime deployment of these addons. There may well be potential alternatives down the line that hook into the larger cluster managers like CoreOS/fleet/etcd or Kubernetes or the stuff that's coming to Docker or Rocket or whatever, and once I understand these alternatives a little better, the separation of concerns between the base addons plugin and the implementation may change. (I still don't really understand "where the data lives" in the "instantly spin up a cluster" world of Kubernetes &co.)

## Service layout

At the root there will be a "docker" directory. This allows services to be implemented using Docker alternatives in the future. (Other aspects can use Docker alternatives just by replacing the plugin, for instance the nginx plugin that can be used instead of nginx-container.)

Under the docker directory:

- an update script, for updating a Docker container to a newer tag.
- A file describing how to link when the service is used as an addon

## Example app addon directory layout

```
PLUSHU_ROOT/apps
`-myapp
  `-addons
    |-rethinkdb
    | |-volume
    | `-container.cid
    `-redis
      |-volume
      `-container.cid
       
```

Config stuff will usually go in the volume, I think. Maybe the addon's directory'll also have a config for args, we'll see how impl shakes out.

Should this maybe be more versatile? Could it be same for an app to have more than one addon of the same service? (I'd say "no" - once that sort of thing is happening there's probably a better way to model that, like even just instances-of-services-that-are-not-addons.)

**Comment by stuartpb** [comment]
One thing I had in the Dokku rethinkdb plugin was a RUNAS_USER that ran the docker container as the `dokku` user. Knowing a bit more now about dokku and Docker, that doesn't make _any_ sense to me (the `dokku` user wouldn't even exist inside the container!) - I think I just cargo-culted that from someone who didn't understand it well either. If the container's supposed to be run as a user _inside_, that'd be handled as part of the Dockerfile.

_(edit 2015-05-29: [It has its purpose.](https://github.com/plushu/plushu-services-docker/issues/2))_

**Comment by stuartpb** [comment]
The saving-static-port-bindings-directly-to-the-host functionality of the dokku-rethinkdb plugin was kind of neat, but right now I'm not really sure I want to integrate that. That might be something I design a whole other set of plugins around in the future. (Like, hey, maybe it can hook into exposing ports from standalone service instances, or addons, or app containers for stuff like Node debugging? And maybe you _don't want to allow_ arbitrary control of your host's ports?)

_(Addendum 2015-06-08: The sort of thing you'd use port exposure for is better handled by SSH tunneling with something like [immpaactt](https://github.com/plushu/immpaactt).)_

**Comment by stuartpb** [comment]
I'm going back and forth for a bit over how update should be implemented. Right now, I'm thinking it should be a base file in the service that handles updating for Docker and otherwise, as in the event of multiple concurrent implementations, it has to control updating the _shared_ components only _once_. (I think having the root update call out to sub-dir updates makes sense.)

In any case, it's important that all service scripts like `update` take a _directory_, so they can be seamlessly used for addons or standalone instances.

_(Edit 2015-05-29: Note, I'm pretty sure I meant "upgrade" and not "update".)_

**Comment by stuartpb** [comment]
I think an alternative to addons-docker might be addons-host, which can hook into systemd/upstart to launch the service as installed on the host. I'm not going to write it, but someone could.

**Comment by stuartpb** [comment]
Should creation logic be left to implementations? Maybe there's some initialization that needs to happen, which would use the Docker image.

Maybe we'll have both the main plugin and implementation plugins listen on add-addon, with the implementations listening between initialization and main "create" being executed. This way, addons can implement both ways and figure it out from there.

**Comment by stuartpb** [comment]
So, addons and services are going to need two separate implementation plugins. services-docker handles the core service mangement hooks (pulling the default image, installing a service, starting a service from a given directory), while addons-docker handles calling them (on app startup) and hooking services into apps.

**Comment by stuartpb** [comment]
I'd say this is pretty well settled and implemented, plushu/plushu-services-docker#2 notwithstanding.
